File: sdks/python/src/echo_agent_sdk/thinking_profile_values.py
from __future__ import annotations

from dataclasses import dataclass

from .thinking import ThinkingLevel
from .thinking_protocol_values import ThinkingProtocol


@dataclass(frozen=True)
class ThinkingProfile:
    protocol: ThinkingProtocol
    levels: tuple[ThinkingLevel, ...]

    @classmethod
    def new(
        cls, protocol: ThinkingProtocol, levels: tuple[ThinkingLevel, ...]
    ) -> ThinkingProfile:
        return cls(protocol, tuple(levels))

    @classmethod
    def unknown(cls) -> ThinkingProfile:
        return cls(ThinkingProtocol.NONE, ())

    def supports_manual_control(self) -> bool:
        return self.protocol.emits_field() and bool(self.levels)


def _version(model: str, prefix: str) -> tuple[int, int] | None:
    if not model.startswith(prefix):
        return None
    segments = model[len(prefix) :].split("-")
    for index, segment in enumerate(segments):
        if "." in segment:
            major_text, minor_text = segment.split(".", 1)
            if "." in minor_text:
                continue
            try:
                major = int(major_text)
                minor = int(minor_text)
            except ValueError:
                continue
        else:
            try:
                major = int(segment)
            except ValueError:
                continue
            minor = 0
            if index + 1 < len(segments):
                try:
                    candidate = int(segments[index + 1])
                except ValueError:
                    candidate = 0
                if 0 <= candidate <= 9:
                    minor = candidate
        if 3 <= major <= 9:
            return major, minor
    return None


def resolve_thinking_profile(
    provider: str,
    model_name: str,
    api_protocol: str = "chat_completions",
    endpoint: str | None = None,
) -> ThinkingProfile:
    provider_lower = provider.strip().lower()
    model = model_name.strip().lower()
    endpoint_lower = (endpoint or "").lower()
    dashscope = (
        provider_lower
        in {"dashscope", "qwen", "aliyun", "alibaba", "modelstudio", "bailian"}
        or "dashscope.aliyuncs.com" in endpoint_lower
    )
    ollama = (
        provider_lower == "ollama"
        or "localhost:11434" in endpoint_lower
        or "127.0.0.1:11434" in endpoint_lower
    )
    if model.startswith("claude-"):
        parsed = _version(model, "claude-")
        if parsed is None or parsed < (4, 6):
            return ThinkingProfile.unknown()
        if parsed == (4, 6):
            return ThinkingProfile.new(
                ThinkingProtocol.ANTHROPIC_EFFORT
                if api_protocol == "anthropic"
                else ThinkingProtocol.OPENAI_REASONING_EFFORT,
                (
                    ThinkingLevel.LOW,
                    ThinkingLevel.MEDIUM,
                    ThinkingLevel.HIGH,
                    ThinkingLevel.XHIGH,
                    ThinkingLevel.MAX,
                ),
            )
        return ThinkingProfile.new(ThinkingProtocol.ANTHROPIC_ADAPTIVE, ())
    if api_protocol == "anthropic":
        return ThinkingProfile.unknown()
    if ollama and api_protocol == "chat_completions":
        if model.startswith("gpt-oss"):
            return ThinkingProfile.new(
                ThinkingProtocol.OLLAMA_THINK,
                (ThinkingLevel.LOW, ThinkingLevel.MEDIUM, ThinkingLevel.HIGH),
            )
        if any(
            model.startswith(prefix)
            for prefix in (
                "qwen3",
                "deepseek-r1",
                "deepseek-v3",
                "deepseek-v4",
                "magistral",
            )
        ):
            return ThinkingProfile.new(
                ThinkingProtocol.OLLAMA_THINK, (ThinkingLevel.NONE, ThinkingLevel.HIGH)
            )
        return ThinkingProfile.unknown()
    if model.startswith(("gpt-5.6", "gpt-5-6")):
        return ThinkingProfile.new(
            ThinkingProtocol.OPENAI_REASONING_EFFORT,
            (
                ThinkingLevel.NONE,
                ThinkingLevel.LOW,
                ThinkingLevel.MEDIUM,
                ThinkingLevel.HIGH,
                ThinkingLevel.XHIGH,
                ThinkingLevel.MAX,
            ),
        )
    if model.startswith("deepseek-v4"):
        return (
            ThinkingProfile.new(
                ThinkingProtocol.ENABLE_THINKING_FLAG,
                (ThinkingLevel.NONE, ThinkingLevel.HIGH),
            )
            if dashscope and api_protocol == "chat_completions"
            else ThinkingProfile.new(
                ThinkingProtocol.DEEPSEEK_REASONING_EFFORT,
                (
                    ThinkingLevel.NONE,
                    ThinkingLevel.LOW,
                    ThinkingLevel.HIGH,
                    ThinkingLevel.MAX,
                ),
            )
        )
    glm = _version(model, "glm-")
    if glm and (glm[0] > 5 or glm >= (5, 2)) and api_protocol == "chat_completions":
        return ThinkingProfile.new(
            ThinkingProtocol.GLM_REASONING_EFFORT,
            (ThinkingLevel.NONE, ThinkingLevel.HIGH, ThinkingLevel.MAX),
        )
    if model.startswith("kimi-k3") and api_protocol == "chat_completions":
        return ThinkingProfile.new(
            ThinkingProtocol.OPENAI_REASONING_EFFORT,
            (ThinkingLevel.LOW, ThinkingLevel.HIGH, ThinkingLevel.MAX),
        )
    if model.startswith("kimi-k2.7"):
        return ThinkingProfile.new(ThinkingProtocol.MODEL_MANAGED, ())
    if model.startswith("kimi-k2.6") and api_protocol == "chat_completions":
        return ThinkingProfile.new(
            ThinkingProtocol.THINKING_TYPE, (ThinkingLevel.NONE, ThinkingLevel.HIGH)
        )
    if model.startswith("qwen3") and api_protocol == "chat_completions":
        return ThinkingProfile.new(
            ThinkingProtocol.ENABLE_THINKING_FLAG,
            (ThinkingLevel.NONE, ThinkingLevel.HIGH),
        )
    if model.startswith("gemini-3") and api_protocol == "chat_completions":
        return ThinkingProfile.new(
            ThinkingProtocol.OPENAI_REASONING_EFFORT,
            (
                ThinkingLevel.MINIMAL,
                ThinkingLevel.LOW,
                ThinkingLevel.MEDIUM,
                ThinkingLevel.HIGH,
            ),
        )
    if model.startswith("gemini-2.5") and api_protocol == "chat_completions":
        return ThinkingProfile.new(
            ThinkingProtocol.OPENAI_REASONING_EFFORT,
            (
                ThinkingLevel.NONE,
                ThinkingLevel.LOW,
                ThinkingLevel.MEDIUM,
                ThinkingLevel.HIGH,
            ),
        )
    return ThinkingProfile.unknown()

File: sdks/typescript/src/thinking_profile_values.ts
import { ThinkingLevel } from "./thinking.js";
import { ThinkingProtocol, thinkingProtocolEmitsField } from "./thinking_protocol_values.js";

export type ThinkingProfile = Readonly<{
  protocol: ThinkingProtocol;
  levels: readonly ThinkingLevel[];
  supportsManualControl(): boolean;
}>;

const GPT_56 = [ThinkingLevel.None, ThinkingLevel.Low, ThinkingLevel.Medium, ThinkingLevel.High, ThinkingLevel.Xhigh, ThinkingLevel.Max] as const;
const CLAUDE_46 = [ThinkingLevel.Low, ThinkingLevel.Medium, ThinkingLevel.High, ThinkingLevel.Xhigh, ThinkingLevel.Max] as const;
const DEEPSEEK_V4 = [ThinkingLevel.None, ThinkingLevel.Low, ThinkingLevel.High, ThinkingLevel.Max] as const;
const GLM_52 = [ThinkingLevel.None, ThinkingLevel.High, ThinkingLevel.Max] as const;
const KIMI_K3 = [ThinkingLevel.Low, ThinkingLevel.High, ThinkingLevel.Max] as const;
const GEMINI_3 = [ThinkingLevel.Minimal, ThinkingLevel.Low, ThinkingLevel.Medium, ThinkingLevel.High] as const;
const GEMINI_25 = [ThinkingLevel.None, ThinkingLevel.Low, ThinkingLevel.Medium, ThinkingLevel.High] as const;
const TOGGLE = [ThinkingLevel.None, ThinkingLevel.High] as const;
const OLLAMA_GPT_OSS = [ThinkingLevel.Low, ThinkingLevel.Medium, ThinkingLevel.High] as const;

function profile(protocol: ThinkingProtocol, levels: readonly ThinkingLevel[]): ThinkingProfile {
  return Object.freeze({
    protocol,
    levels: Object.freeze([...levels]),
    supportsManualControl: () => thinkingProtocolEmitsField(protocol) && levels.length > 0,
  });
}

function version(model: string, prefix: string): [number, number] | undefined {
  const rest = model.startsWith(prefix) ? model.slice(prefix.length) : "";
  const segments = rest.split("-");
  for (let index = 0; index < segments.length; index += 1) {
    const segment = segments[index] ?? "";
    const match = /^(\d+)(?:\.(\d+))?$/u.exec(segment);
    if (!match) continue;
    const major = Number(match[1]);
    const next = segments[index + 1];
    const hyphenMinor = match[2] === undefined && next !== undefined && /^\d+$/u.test(next) && Number(next) <= 9 ? next : "0";
    const minor = Number(match[2] ?? hyphenMinor);
    if (major >= 3 && major <= 9) return [major, minor];
  }
  return undefined;
}

/** Resolve the provider/model thinking dialect without contacting a provider. */
export function resolveThinkingProfile(provider: string, modelName: string, apiProtocol = "chat_completions", endpoint?: string): ThinkingProfile {
  const providerLower = provider.trim().toLowerCase();
  const model = modelName.trim().toLowerCase();
  const endpointLower = (endpoint ?? "").toLowerCase();
  const dashscope = ["dashscope", "qwen", "aliyun", "alibaba", "modelstudio", "bailian"].includes(providerLower) || endpointLower.includes("dashscope.aliyuncs.com");
  const ollama = providerLower === "ollama" || endpointLower.includes("localhost:11434") || endpointLower.includes("127.0.0.1:11434");
  if (model.startsWith("claude-")) {
    const parsed = version(model, "claude-");
    if (!parsed || parsed[0] < 4 || parsed[0] === 4 && parsed[1] < 6) return profile(ThinkingProtocol.None, []);
    if (parsed[0] === 4 && parsed[1] === 6) return profile(apiProtocol === "anthropic" ? ThinkingProtocol.AnthropicEffort : ThinkingProtocol.OpenaiReasoningEffort, CLAUDE_46);
    return profile(ThinkingProtocol.AnthropicAdaptive, []);
  }
  if (apiProtocol === "anthropic") return profile(ThinkingProtocol.None, []);
  if (ollama && apiProtocol === "chat_completions") {
    if (model.startsWith("gpt-oss")) return profile(ThinkingProtocol.OllamaThink, OLLAMA_GPT_OSS);
    if (["qwen3", "deepseek-r1", "deepseek-v3", "deepseek-v4", "magistral"].some((prefix) => model.startsWith(prefix))) return profile(ThinkingProtocol.OllamaThink, TOGGLE);
    return profile(ThinkingProtocol.None, []);
  }
  if (model.startsWith("gpt-5.6") || model.startsWith("gpt-5-6")) return profile(ThinkingProtocol.OpenaiReasoningEffort, GPT_56);
  if (model.startsWith("deepseek-v4")) return profile(dashscope && apiProtocol === "chat_completions" ? ThinkingProtocol.EnableThinkingFlag : ThinkingProtocol.DeepseekReasoningEffort, dashscope && apiProtocol === "chat_completions" ? TOGGLE : DEEPSEEK_V4);
  const glm = version(model, "glm-");
  if (glm && (glm[0] > 5 || glm[0] === 5 && glm[1] >= 2) && apiProtocol === "chat_completions") return profile(ThinkingProtocol.GlmReasoningEffort, GLM_52);
  if (model.startsWith("kimi-k3") && apiProtocol === "chat_completions") return profile(ThinkingProtocol.OpenaiReasoningEffort, KIMI_K3);
  if (model.startsWith("kimi-k2.7")) return profile(ThinkingProtocol.ModelManaged, []);
  if (model.startsWith("kimi-k2.6") && apiProtocol === "chat_completions") return profile(ThinkingProtocol.ThinkingType, TOGGLE);
  if (model.startsWith("qwen3") && apiProtocol === "chat_completions") return profile(ThinkingProtocol.EnableThinkingFlag, TOGGLE);
  if (model.startsWith("gemini-3") && apiProtocol === "chat_completions") return profile(ThinkingProtocol.OpenaiReasoningEffort, GEMINI_3);
  if (model.startsWith("gemini-2.5") && apiProtocol === "chat_completions") return profile(ThinkingProtocol.OpenaiReasoningEffort, GEMINI_25);
  return profile(ThinkingProtocol.None, []);
}

export const ThinkingProfile = Object.freeze({
  new(protocol: ThinkingProtocol, levels: readonly ThinkingLevel[]): ThinkingProfile { return profile(protocol, levels); },
  unknown(): ThinkingProfile { return profile(ThinkingProtocol.None, []); },
});
